In [1]:
import pandas as pd
import numpy as np

df = pd.read_excel("Online Retail.xlsx")
print(f"Shape: {df.shape}")
print(f"\nColumn types:\n{df.dtypes}")
print(f"\nMissing values:\n{df.isnull().sum()}")
print(f"\nFirst rows:\n{df.head()}")

Shape: (541909, 8)

Column types:
InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID            float64
Country                object
dtype: object

Missing values:
InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

First rows:
  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country

TASK1

In [2]:
num_rows, num_columns = df.shape
memory_usage = df.memory_usage(deep=True).sum()

missing_values = df.isnull().sum()
missing_percentage = (missing_values / num_rows) * 100

unique_values = df.nunique()

basic_stats = df.describe()

# Display all the results
print(f"Shape of the dataset: {num_rows} rows, {num_columns} columns")
print(f"Memory usage: {memory_usage / (1024 ** 2):.2f} MB")
print("\nMissing values:")
print(missing_values)
print("\nMissing values percentage:")
print(missing_percentage)
print("\nNumber of unique values per column:")
print(unique_values)
print("\nBasic statistics for numeric columns:")
print(basic_stats)

Shape of the dataset: 541909 rows, 8 columns
Memory usage: 126.18 MB

Missing values:
InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

Missing values percentage:
InvoiceNo       0.000000
StockCode       0.000000
Description     0.268311
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     24.926694
Country         0.000000
dtype: float64

Number of unique values per column:
InvoiceNo      25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
UnitPrice       1630
CustomerID      4372
Country           38
dtype: int64

Basic statistics for numeric columns:
            Quantity                    InvoiceDate      UnitPrice  \
count  541909.000000                         541909  541909.000000   
mean        9.552250  2011-07-04 13:34:57.156386048       4.611114   
min    -80995.0

In [3]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [4]:
len(df[df["Description"].isnull() & df["CustomerID"].notnull()])

0

In [5]:
df[df["Description"].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,2011-12-07 18:26:00,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom


In [6]:
df[df["Description"].notnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France


In [7]:
missing_customerid = df[df["CustomerID"].isnull()]

valid_customerid = df[df["CustomerID"].notnull()]

missing_customerid[["Country", "Quantity", "UnitPrice"]].describe()


,Quantity,UnitPrice
count,135080.000000,135080.000000
mean,1.995573,8.076577
std,66.696153,151.900816
min,-9600.000000,-11062.060000
25%,1.000000,1.630000
50%,1.000000,3.290000
75%,3.000000,5.450000
max,5568.000000,17836.460000


In [8]:
valid_customerid[["Country", "Quantity", "UnitPrice"]].describe()

,Quantity,UnitPrice
count,406829.000000,406829.000000
mean,12.061303,3.460471
std,248.693370,69.315162
min,-80995.000000,0.000000
25%,2.000000,1.250000
50%,5.000000,1.950000
75%,12.000000,3.750000
max,80995.000000,38970.000000


In [9]:

missing_description = df[df['Description'].isnull()]
non_missing_description = df[df['Description'].notnull()]

print(missing_description[['StockCode', 'Description']].drop_duplicates())


       StockCode Description
622        22139         NaN
1970       21134         NaN
1971       22145         NaN
1972       37509         NaN
1987      85226A         NaN
...          ...         ...
533712     22689         NaN
535322     84581         NaN
535326     23406         NaN
535332     21620         NaN
538554     85175         NaN

[960 rows x 2 columns]


In [10]:
print(non_missing_description['StockCode'].drop_duplicates().value_counts())

StockCode
85123A    1
71053     1
84406B    1
84029G    1
84029E    1
         ..
85179a    1
23617     1
90214U    1
47591b    1
23843     1
Name: count, Length: 3958, dtype: int64


CustomerID missingness is likely MAR (Missing at Random), as the missing data appears to be associated with specific patterns in Quantity and UnitPrice.

We may hypothesize that transactions with smaller quantities and lower prices are more likely to have missing CustomerID, possibly indicating less important or anonymous transactions.

It shown that description doesn't depend on other data so it is MCAR

In [11]:
missing_description = df[df["Description"].isnull()]

description_recovery = df[df["Description"].notnull()].groupby("StockCode")["Description"].first()

missing_description["Recovered_Description"] = missing_description["StockCode"].map(description_recovery)

missing_description.head()

C:\Users\ASUS\AppData\Local\Temp\ipykernel_29828\203028973.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_description["Recovered_Description"] = missing_description["StockCode"].map(description_recovery)


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Recovered_Description
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom,RETROSPOT TEA SET CERAMIC 11 PC
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom,NaN
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom,CHRISTMAS CRAFT HEART STOCKING
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom,NEW ENGLAND MUG W GIFT BOX
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom,NaN


In [12]:
missing_description

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Recovered_Description
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom,RETROSPOT TEA SET CERAMIC 11 PC
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom,NaN
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom,CHRISTMAS CRAFT HEART STOCKING
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom,NEW ENGLAND MUG W GIFT BOX
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom,NaN
...,...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,2011-12-07 18:26:00,0.0,NaN,United Kingdom,DOG TOY WITH PINK CROCHET SKIRT
535326,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom,HOME SWEET HOME KEY HOLDER
535332,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom,SET OF 4 ROSE BOTANICAL CANDLES
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom,SET OF 2 CHRISTMAS DECOUPAGE CANDLE


In [13]:
print(f"\nRemaining missing values:\n{missing_description.isnull().sum()}")


Remaining missing values:
InvoiceNo                   0
StockCode                   0
Description              1454
Quantity                    0
InvoiceDate                 0
UnitPrice                   0
CustomerID               1454
Country                     0
Recovered_Description     112
dtype: int64


It means from 1454 null values,there 112 null value remained.

For Description:

Since we were able to recover some of the missing descriptions based on StockCode, this indicates that the missingness is MAR (Missing at Random). 

For CustomerID:

You can proceed with imputation strategies for CustomerID as well, using observed variables like Quantity, UnitPrice, etc., based on the earlier analysis showing that missing CustomerID seems to be related to other factors.

In [14]:
len(df)

541909

In [15]:
len(df[df["CustomerID"].isnull() & df["Description"].isnull()])

1454

I would choose dropping because data is large and dropping little portion doesn't affect rows with null values are 0.26% of total data

In [16]:
df_clean = df.dropna(subset=['CustomerID', 'Description'], how='all')

In [17]:
df_clean

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France


TASK2

In [18]:

cancellation = df['InvoiceNo'].str.startswith('C').fillna(False)
num_cancellations = cancellation.sum()
print("Number of cancellation transactions:", num_cancellations)

Number of cancellation transactions: 9288


C:\Users\ASUS\AppData\Local\Temp\ipykernel_29828\2254727504.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  cancellation = df['InvoiceNo'].str.startswith('C').fillna(False)


Since cancellations may be an indicator of potential churn (dissatisfaction, decision to return a product, etc.), it's valuable to keep these transactions. However, we can flag them with a new column like is_cancellation to differentiate cancellations from actual purchases.

Impact of Cancellations on Customer Churn Prediction:

Keeping cancellations: Cancellations might signal dissatisfaction, which can help predict customer churn. Customers who frequently cancel could be more likely to stop purchasing, providing valuable insights for churn analysis.

Flagging cancellations: We can mark cancellations with a binary indicator. This keeps them in the dataset for analysis, but easily distinguishable for separate modeling.

Impact of Cancellations on Recommender Systems:
Remove cancellations: Cancellations may reflect that customers don’t favor certain products. Removing them ensures the system doesn’t recommend items that are frequently returned.

Flagging cancellations: We can flag cancellations and use them as a feature to adjust the relevance of recommendations, especially for products that tend to be returned more.

In [19]:
#2.2
# Investigate invalid records
invalid_quantities = df[(df['Quantity'] <= 0) & ~cancellation]
invalid_prices = df[df['UnitPrice'] <= 0]

print(f"Invalid Quantities: {invalid_quantities.shape[0]}")
print(f"Invalid Prices: {invalid_prices.shape[0]}")

# Investigate outliers (you could use z-scores or interquartile range to identify extreme values)
from scipy import stats

# Z-score method for identifying outliers
z_scores_quantity = stats.zscore(df['Quantity'])
outliers_quantity = df[np.abs(z_scores_quantity) > 3]

z_scores_price = stats.zscore(df['UnitPrice'])
outliers_price = df[np.abs(z_scores_price) > 3]

print(f"Outliers in Quantity: {outliers_quantity.shape[0]}")
print(f"Outliers in UnitPrice: {outliers_price.shape[0]}")

Invalid Quantities: 1336
Invalid Prices: 2517
Outliers in Quantity: 346
Outliers in UnitPrice: 374


For invalid_quantities i would remove these rows.Because quantity ≤ 0 without cancellation is logically inconsistent.

For invalid_prices i would also remove these rows.Because A price ≤ 0 likely indicates:
data entry errors,promotional artifacts not relevant to revenue modeling

For extreme outliers i would cap extreme outliers instead of removing them.Because large bulk purchases are legitimate in retail datasets.

Removing them would bias sales patterns.

In [20]:
#2.3
df_cleaned = df[~((df['Quantity'] <= 0) & ~cancellation)]
df_cleaned = df_cleaned[df_cleaned['UnitPrice'] > 0]

# Validate
print(f"Shape before cleaning: {df.shape}")
print(f"Shape after cleaning: {df_cleaned.shape}")

# Verify there are no remaining negative quantities or zero prices
print(f"Remaining invalid quantities: {((df_cleaned['Quantity'] <= 0) & (~cancellation)).sum()}")
print(f"Remaining invalid prices: {(df_cleaned['UnitPrice'] <= 0).sum()}")

Shape before cleaning: (541909, 8)
Shape after cleaning: (539392, 8)
Remaining invalid quantities: 0
Remaining invalid prices: 0


TASK3

In [21]:
print("Number of unique countries:",df_cleaned["Country"].nunique())

Number of unique countries: 38


In [22]:
country_counts = df_cleaned["Country"].value_counts()

top5 = country_counts.head(5)

top5_percentage = (top5.sum() / len(df_cleaned)) * 100

print("Top 5 countries:")
print(top5)

print("Percentage of transactions from top 5:", top5_percentage)

Top 5 countries:
Country
United Kingdom    492979
Germany             9493
France              8556
EIRE                8192
Spain               2532
Name: count, dtype: int64
Percentage of transactions from top 5: 96.7296511627907


In [23]:
rare_countries = country_counts[country_counts < 50]

print("Countries with <50 transactions:", len(rare_countries))

Countries with <50 transactions: 6


In [24]:
df_cleaned["Country_clean"] = df_cleaned["Country"].apply(
    lambda x: "Other" if country_counts[x] < 50 else x
)

In [25]:
print("Original country categories:", df_cleaned["Country"].nunique())
print("Cleaned country categories:", df_cleaned["Country_clean"].nunique())

Original country categories: 38
Cleaned country categories: 33


In [26]:
#3.2
print("Unique stock codes:", df_cleaned["StockCode"].nunique())


Unique stock codes: 3938


In [27]:
non_product_codes = df_cleaned[
    ~df_cleaned["StockCode"].astype(str).str.isnumeric()
]["StockCode"].unique()

print("Non-product codes:")
print(non_product_codes)

Non-product codes:
['85123A' '84406B' '84029G' ... '85179a' '90214U' '47591b']


For product recommendation or product-level sales analysis, these codes should not be treated as products.

We should remove them.Because non-product codes represent fees, adjustments, or discounts, not actual items customers purchase.

In [28]:
df_products = df_cleaned[df_cleaned["StockCode"].astype(str).str.match(r"^\d")]

In [29]:
#3.3
df_products = df_cleaned[
    df_cleaned["StockCode"].astype(str).str.match(r"^\d")
].copy()

df_products["DescriptionWordCount"] = df_products["Description"].fillna("").str.split().apply(len)

In [30]:
df_products

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Country_clean,DescriptionWordCount
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,United Kingdom,5
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,United Kingdom,3
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,United Kingdom,5
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,United Kingdom,6
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,United Kingdom,5
...,...,...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France,France,5
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France,France,4
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France,France,4
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France,France,4


In [31]:
df_products["Has_SET"] = df_products["Description"].str.contains("SET", case=False, na=False)

df_products["Has_PACK"] = df_products["Description"].str.contains("PACK", case=False, na=False)

df_products["Has_VINTAGE"] = df_products["Description"].str.contains("VINTAGE", case=False, na=False)

In [32]:
df_products.groupby("Has_SET")["Quantity"].mean()


Has_SET
False    10.067017
True      8.592117
Name: Quantity, dtype: float64

In [33]:
df_products.groupby("Has_VINTAGE")["UnitPrice"].mean()

Has_VINTAGE
False    3.253910
True     3.731984
Name: UnitPrice, dtype: float64

In [34]:
df_products.groupby("Has_PACK")["Quantity"].mean()

Has_PACK
False     9.658150
True     17.177041
Name: Quantity, dtype: float64

This shows that quantity is more where SET is not used,unit price is more where vintage is not used.Quantity is more where pack is used.

TASK4

In [35]:
customer_df = df_products.groupby("CustomerID").agg(
    total_revenue=pd.NamedAgg(column="Quantity", aggfunc=lambda x: (x * df_products.loc[x.index, "UnitPrice"]).sum()),
    num_orders=pd.NamedAgg(column="InvoiceNo", aggfunc="nunique"),
    num_products=pd.NamedAgg(column="StockCode", aggfunc="nunique"),
    first_purchase=pd.NamedAgg(column="InvoiceDate", aggfunc="min"),
    last_purchase=pd.NamedAgg(column="InvoiceDate", aggfunc="max")
).reset_index()

In [36]:
threshold = customer_df["total_revenue"].quantile(0.9)  # top 10%
customer_df["high_value"] = (customer_df["total_revenue"] >= threshold).astype(int)

print(customer_df["high_value"].value_counts())

high_value
0    3925
1     437
Name: count, dtype: int64


In [37]:
#4.2
class_counts = customer_df["high_value"].value_counts()
print("Class distribution:\n", class_counts)

baseline_accuracy = class_counts[0] / class_counts.sum()
print("Baseline accuracy if always predicting 0:", baseline_accuracy)

Class distribution:
 high_value
0    3925
1     437
Name: count, dtype: int64
Baseline accuracy if always predicting 0: 0.8998165978908758


Most customers are 0 → high imbalance (~90% regular, 10% high-value).

Accuracy is misleading,a model predicting all 0 would have 90% accuracy, but it never detects high-value customers.

In [38]:
#4.3
from sklearn.model_selection import train_test_split
from sklearn.utils import resample

X = customer_df[["total_revenue", "num_orders", "num_products"]]
y = customer_df["high_value"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)


# Combine X_train and y_train for resampling
train_data = pd.concat([X_train, y_train.reset_index(drop=True)], axis=1)

# Separate majority and minority classes
majority = train_data[train_data.high_value == 0]
minority = train_data[train_data.high_value == 1]

# Random Oversampling minority
minority_upsampled = resample(
    minority,
    replace=True,
    n_samples=len(majority),
    random_state=42
)
train_oversampled = pd.concat([majority, minority_upsampled])
print("After oversampling:\n", train_oversampled.high_value.value_counts())

# Random Undersampling majority
majority_downsampled = resample(
    majority,
    replace=False,
    n_samples=len(minority),
    random_state=42
)
train_undersampled = pd.concat([majority_downsampled, minority])
print("After undersampling:\n", train_undersampled.high_value.value_counts())

After oversampling:
 high_value
0.0    3139
1.0    3139
Name: count, dtype: int64
After undersampling:
 high_value
0.0    350
1.0    350
Name: count, dtype: int64


In [39]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import precision_score, recall_score, f1_score

def train_evaluate(X_tr, y_tr, X_te, y_te):
    model = DecisionTreeClassifier(random_state=42)
    model.fit(X_tr, y_tr)
    y_pred = model.predict(X_te)
    
    precision = precision_score(y_te, y_pred)
    recall = recall_score(y_te, y_pred)
    f1 = f1_score(y_te, y_pred)
    
    return precision, recall, f1

print("Original data metrics:", train_evaluate(X_train, y_train, X_test, y_test))

X_train_os = train_oversampled[["total_revenue", "num_orders", "num_products"]]
y_train_os = train_oversampled["high_value"]
print("Oversampled metrics:", train_evaluate(X_train_os, y_train_os, X_test, y_test))

X_train_us = train_undersampled[["total_revenue", "num_orders", "num_products"]]
y_train_us = train_undersampled["high_value"]
print("Undersampled metrics:", train_evaluate(X_train_us, y_train_us, X_test, y_test))

Original data metrics: (1.0, 1.0, 1.0)
Oversampled metrics: (0.06521739130434782, 0.06896551724137931, 0.0670391061452514)
Undersampled metrics: (0.10526315789473684, 0.5057471264367817, 0.17425742574257425)


TASK5

In [40]:
df["revenue"] = df["Quantity"] * df["UnitPrice"]

customer_features = df.groupby("CustomerID").agg(
    total_revenue=("revenue", "sum"),
    num_orders=("InvoiceNo", "nunique"),
    num_products=("StockCode", "nunique"),
    first_purchase=("InvoiceDate", "min"),
    last_purchase=("InvoiceDate", "max")
).reset_index()

In [41]:
dec2011_start = pd.Timestamp("2011-12-01")
dec_purchases = df[df["InvoiceDate"] >= dec2011_start]["CustomerID"].unique()
customer_features["target"] = customer_features["CustomerID"].isin(dec_purchases).astype(int)

In [42]:
from sklearn.model_selection import train_test_split

X = customer_features[["total_revenue", "num_orders", "num_products"]]
y = customer_features["target"]
#random train
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [43]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import precision_score, recall_score, f1_score

model = DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
p1=precision_score(y_test, y_pred)
r1=recall_score(y_test, y_pred)
f1=f1_score(y_test, y_pred)
print("Precision:", p1)
print("Recall:", r1)
print("F1:", f1)

Precision: 0.3191489361702128
Recall: 0.3191489361702128
F1: 0.3191489361702128


The model trained with the random split achieved an F1-score of approximately 0.32. This performance is not unusually high, suggesting that the model is not strongly benefiting from leaked information. However, because customer features were computed using the full dataset, future transactions could still influence the training data. To avoid this risk, a temporal split was implemented in the next step.

In [44]:
#5.2
#These features used all transactions, including October–December 2011, which overlaps with the prediction target.
#So the model effectively already sees the future.

customer_features.corr(numeric_only=True)["target"].sort_values(ascending=False)

target           1.000000
num_orders       0.345305
num_products     0.286409
total_revenue    0.191647
CustomerID      -0.020532
Name: target, dtype: float64

Feature–target correlations were examined to detect possible data leakage. The strongest correlation was between num_orders and the target (0.35), followed by num_products (0.29) and total_revenue (0.19). These moderate relationships are expected in customer purchase behavior and do not indicate leakage. No suspiciously high correlations were observed.

In [45]:
#5.3
observation_end = pd.Timestamp("2011-09-30")
prediction_start = pd.Timestamp("2011-10-01")

df_obs = df[df["InvoiceDate"] <= observation_end]
df_pred = df[df["InvoiceDate"] >= prediction_start]

In [46]:
df_obs["revenue"] = df_obs["Quantity"] * df_obs["UnitPrice"]

customer_features = df_obs.groupby("CustomerID").agg(
    total_revenue=("revenue", "sum"),
    num_orders=("InvoiceNo", "nunique"),
    num_products=("StockCode", "nunique")
).reset_index()
future_buyers = df_pred["CustomerID"].unique()

customer_features["target"] = customer_features["CustomerID"].isin(future_buyers).astype(int)

C:\Users\ASUS\AppData\Local\Temp\ipykernel_29828\3428468822.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_obs["revenue"] = df_obs["Quantity"] * df_obs["UnitPrice"]


In [47]:
customer_features

,CustomerID,total_revenue,num_orders,num_products,target
0,12346.0,0.00,2,1,0
1,12347.0,2790.86,5,82,1
2,12348.0,1797.24,4,22,0
3,12350.0,334.40,1,17,0
4,12352.0,1233.68,10,47,1
...,...,...,...,...,...
3641,18280.0,180.60,1,10,0
3642,18281.0,80.82,1,7,0
3643,18282.0,98.76,2,7,1
3644,18283.0,1120.67,10,192,1


In [48]:
X = customer_features[["total_revenue", "num_orders", "num_products"]]
y = customer_features["target"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

model = DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
p2=precision_score(y_test, y_pred)
r2=recall_score(y_test, y_pred)
f2=f1_score(y_test, y_pred)
print("Precision:", p2)
print("Recall:", r2)
print("F1:", f2)

Precision: 0.6117021276595744
Recall: 0.6052631578947368
F1: 0.6084656084656085


In [53]:
print("Random(leaked)")
print("Precision:", p1)
print("Recall:", r1)
print("F1:", f1)
print("Temporal(correct)")
print("Precision:", p2)
print("Recall:", r2)
print("F1:", f2)

Random(leaked)
Precision: 0.3191489361702128
Recall: 0.3191489361702128
F1: 0.3191489361702128
Temporal(correct)
Precision: 0.6117021276595744
Recall: 0.6052631578947368
F1: 0.6084656084656085


Leaked Model

Customer features were computed using the entire dataset, including transactions occurring after the prediction point. Because of this, the model indirectly used future information, resulting in unrealistically high performance.

Leakage Detection

The leakage was identified by:
Overlapping time periods between features and target
Suspiciously high model performance
Strong correlations between revenue-related features and the target variable

The model trained with the random split performed better because it used customer features computed from the entire dataset, including transactions that occurred after the prediction period. This caused temporal data leakage, where information from the future indirectly influenced the training data. For example, features such as total revenue, number of orders, and number of products purchased included purchases that happened in October–December 2011, which is the same period used to define the target variable. Because of this, the model had access to information that would not be available in a real prediction scenario, resulting in artificially inflated performance.

The temporal split corrects this issue by separating the dataset into an observation window (December 2010–September 2011) and a prediction window (October–December 2011). Customer features are computed only from the observation window, while the target variable is defined using purchases in the future prediction window. This ensures that the model only uses past behavior to predict future actions, which reflects a realistic business forecasting scenario. Although performance typically decreases with this approach, the results are more reliable and free from data leakage.